File: src/com/kraftwerking/datastructuresalgo/leetcode/UniquePaths62.java
package com.kraftwerking.datastructuresalgo.leetcode;


public class UniquePaths62 {

  public int uniquePaths(int m, int n) {
    int[] row = new int[n];
    for (int i = 0; i < n; i++) {
      row[i] = 1;
    }

    for (int ii = 0; ii < (m - 1); ii++) {
      int[] newRow = new int[n];
      for (int i = 0; i < n; i++) {
        newRow[i] = 1;
      }

      for (int j = (n - 2); j >= 0; j--) {
        newRow[j] = newRow[j + 1] + row[j];
      }
      row = newRow;
    }
    return row[0];
  }

  public static void main(String[] args) {
    UniquePaths62 uniquePaths62 = new UniquePaths62();
    int result = uniquePaths62.uniquePaths(3, 7);
    System.out.println(result);

  }
}

File: python/leetcode/UniquePaths62.py
from typing import List


class Solution(object):
    def uniquePaths(self, m: int, n: int) -> int:
        row = [1] * n

        for i in range(m - 1):
            newRow = [1] * n
            for j in range(n - 2, -1, -1):
                newRow[j] = newRow[j + 1] + row[j]
            row = newRow
        return row[0]


def main():
    m = 3
    n = 7

    solution = Solution()
    result = solution.uniquePaths(m, n)
    print(result)


if __name__ == "__main__":
    main()
